Add an optional extra level of checking: ASSUME(...) - an opt-in side-effect safe assert(...)

practicalswift opened the pull request:
As suggested by sipa in the open issue #4576 (2014) -- "[discussion] Dealing with assertions and optional consistency checking" -- as an alternative to `assert(…)` in situations where `assert(…)` is not appropriate:

> …
> What I want is more of these checks, more as a way for the programmer to say "this is what I assume here", more than "if this doesn't hold here, we're in BIG trouble". It makes the code clearer, and simultaneously verifies that such assumptions hold. But only in cases where we're not at risk of hurting the network by dying.
> …

`ASSUME(expression)` works like this:
* If compiled with `-DABORT_ON_FAILED_ASSUME` (set by `--enable-debug` and/or `--enable-fuzz`): Evaluate expression and abort if `expression` is `false`.
* If compiled without `-DABORT_ON_FAILED_ASSUME`: Evaluate `expression` and continue execution.

Example:

```
int main(void) {
    ASSUME(IsFoo());
     ...
}
```

If `!IsFoo()` _and_ `-DABORT_ON_FAILED_ASSUME`, then:

```
    filename.cpp:123: main: ASSUME(IsFoo()) failed.
    Aborted
```

Otherwise the execution continues.

Resolves #4576.

ajtowns commented:
If you're running the CHECK no matter what, could consider making the failure configurable at runtime (could default to not abort on mainnet, but to abort on testnet/regtest maybe?). Something like:

    do {
         if (g_abort_on_check_fail) {
              fprintf(stderr, ...); std::abort();
         } else {
              LogPrintf(...);
         }
    } while(0)

practicalswift commented:
@ajtowns I like the idea of making the default depend on the network. I'll let others chime in regarding how to choose the defaults and how to override it and then update based on the feedback :-)

sdaftuar commented:
Concept ACK -- I often want something that's like an assert but wouldn't cause cascading failure on the network if I'm wrong.

morcos commented:
Concept ACK, but I also really like the idea of LogPrinting these (either no matter what or at least if some level of debug logging is on).   Part of the benefit would be in catching rare conditions that might not occur in regular testing but would show up occasionally among thousands of real world users.  The LogPrint could direct the user to report the error on GitHub.

MarcoFalke commented:
The link could even prefill all the additional debug information. E.g. https://github.com/bitcoin/bitcoin/issues/new?title=file_name.cpp:LINE_NR%20CHECK%20failed%20bla&body=more%20bla

Empact commented:
Concept ACK

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -1691,6 +1692,7 @@ bool static ProcessHeadersMessage(CNode *pfrom, CConnman *connman, const std::ve
                 }
                 pindexWalk = pindexWalk->pprev;
             }
+            CHECK(pindexWalk);
```
Looks like you are using this wrong. This should be an assert, since it is dereferenced later on.

practicalswift replied:
I agree it is not an optimal example: `assert(…);` would obviously be the more appropriate choice for cases where we want an unconditional `abort` if the assumption does not hold. I'll remove this example.

MarcoFalke commented on src/check.h:
```diff
@@ -0,0 +1,50 @@
+// * If compiled with -DABORT_ON_FAILED_CHECK (set by --enable-debug):
+//   Evaluate expression and abort if expression is falsy.
+//
+// * If compiled without -DABORT_ON_FAILED_CHECK:
```
Shorter without loss of precision:

```suggestion
// * otherwise:
```

practicalswift replied:
Good point! Fixed!

MarcoFalke commented on src/check.h:
```diff
@@ -0,0 +1,50 @@
+#define CHECK_FAILURE(expression, file, line, func) ((void)0)
+#endif
+
+#define CHECK(expression)                                             \
```
This is missing an explanation when to use it. How is it different from `assert` or `std::logic_error`? Where should it be used in net or in consensus/validation code?

practicalswift replied:
Added additional documentation. Please re-review :-)

maflcko reviewed (commented):
Concept ACK if done right (with proper examples and documentation), but currently it seems to be confusing `assert` with `CHECK`.

And maybe it should be called `ASSUME`?

ryanofsky commented:
This change seems good, but I'd suggest calling the macro `DCHECK` instead of `CHECK` because:

- It would suggest that the macro has different behavior in debug vs release builds.
- It would allow us to create a corresponding `CHECK` macro in the future to replace our current non-standard usages of `assert` (in standard c++, asserts are compiled out of release builds, but not in bitcoin c++).
- It would be more consistent with other support libraries ([glog](https://github.com/google/glog/blob/ba8a9f6952d04d1403b97df24e6836227751454e/src/glog/logging.h.in#L582-L586), [folly](https://github.com/facebook/folly/blob/5a68ed6f9cfbe634af578b573e092a16b038c02e/folly/logging/xlog.h#L387-L395)) which use `CHECK` to abort unconditionally and `DCHECK` to only abort in debug builds.

For reference:

|                                                                                                                                   | Condition is compiled | Condition is evaluated | Program is aborted   |
|-----------------------------------------------------------------------------------------------------------------------------------|-----------------------|------------------------|----------------------|
| C++ standard `assert`                                                                                                             | if `NDEBUG` is unset  | if `NDEBUG` is unset   | if `NDEBUG` is unset |
| Bitcoin `assert`                                                                                                                  | always                | always                 | always               |
| glog [`CHECK`](https://github.com/google/glog/blob/ba8a9f6952d04d1403b97df24e6836227751454e/src/glog/logging.h.in#L582-L586)      | always                | always                 | always               |
| glog [`DCHECK`](https://github.com/google/glog/blob/ba8a9f6952d04d1403b97df24e6836227751454e/src/glog/logging.h.in#L1034-L1038)   | always                | if `NDEBUG` is unset   | if `NDEBUG` is unset |
| folly [`XCHECK`](https://github.com/facebook/folly/blob/5a68ed6f9cfbe634af578b573e092a16b038c02e/folly/logging/xlog.h#L387-L395)  | always                | always                 | always               |
| folly [`XDCHECK`](https://github.com/facebook/folly/blob/5a68ed6f9cfbe634af578b573e092a16b038c02e/folly/logging/xlog.h#L445-L457) | if `NDEBUG` is unset  | if `NDEBUG` is unset   | if `NDEBUG` is unset |

Re: https://github.com/bitcoin/bitcoin/pull/16136#issuecomment-499862751 and https://github.com/bitcoin/bitcoin/pull/16136#issuecomment-499866876 I think what Marco and Alex are asking for is useful but out of scope for an assert or check macro. I think asserts and checks are best for succinctly conveying assumptions made by the code author that in no cases should ever be violated, and conveying at a very coarse level how crucial those assumptions are (abort-worthy or not). The main point of asserts and checks should be to make code more readable and communicate how to think about it.

Once you're in the realm of conditions that you think should never happen, but are more complicated and might end up happening anyway, you are really better off writing an actual log print with text that would put the error into context. You may also want to abort in these cases, which is why logging frameworks support `FATAL` and `DFATAL` severity levels.

practicalswift committed:
Add ASSUME(expr). Set -ABORT_ON_ASSUME_FAIL when configured with --enable-debug or --enable-fuzz.

MarcoFalke commented on src/check.h:
```diff
@@ -0,0 +1,64 @@
+// ==================
+// DCHECK(...) is used to document assumptions.
+//
+// Bitcoin is always compiled with assertions enabled. assert(...)
```
```suggestion
// Bitcoin Core is always compiled with assertions enabled. assert(...)
```

practicalswift replied:
Fixed!

MarcoFalke commented on src/check.h:
```diff
@@ -0,0 +1,64 @@
+#ifdef ABORT_ON_FAILED_DCHECK
+#define DCHECK_FAILURE(expression, file, line, func)       \
+    do {                                                  \
+        fprintf(stderr, "%s:%d: %s: DCHECK(%s) failed.\n", \
```
```suggestion
        tfm::format(std::cerr, "%s:%d: %s: DCHECK(%s) failed.\n", \
```

practicalswift replied:
Changed!

ryanofsky commented on src/check.h:
```diff
@@ -0,0 +1,64 @@
+
+#define DCHECK(expression)                                             \
+    do {                                                              \
+        if (!(expression)) {                                          \
```
It's possible I misunderstood part of the intent of this PR, but I don't think the expression should be evaluated in non-debug builds. Few reasons:

- It would encourage putting code that has external side effects in DCHECK macros, which would make it easier not to notice the effects of invoking code, and be bad for readability.
- It would encourage general misuse of the DCHECK macro. Like I wrote in my earlier comment I think the best use of DCHECK is just to state basic assumptions that code makes so it is more readable, and to help catch simple bugs during development. It shouldn't be used as lazy way to handle errors in release builds and avoid having to write readable error messages that put errors into context.
- It could potentially make release code less efficient.
- It would be inconsistent with other DCHECK implementations. Neither the glog nor folly DCHECK macros have side effects in release builds (see https://github.com/bitcoin/bitcoin/pull/16136#issuecomment-501779934) .

My suggestion would be to rename `ABORT_ON_FAILED_DCHECK` to `ENABLE_DCHECK` and when it's not defined either make `DCHECK` a no-op like folly, or make it compile the condition expression without executing it like glog.

practicalswift replied:
It seemed like the consensus in #4576 was to make the expression evaluated also in non-debug builds. FWIW, that is consistent with how it is done in `secp256k1`.

I don't have a strong opinion here: I'll let others chime in and adjust the code to meet consensus.

ajtowns replied:
Seems like clarifying the intent of the PR would be a good first step...

I think there's four things we could have:

 * `assert(chk)` -- always evaluate, always abort if it failes
 * `CHECK(chk)` -- always evaluate, but only abort if in regtest or similar; maybe log errors if not aborting.
 * `debug_assert(chk)` -- evaluate condition only when invoked with a special option like `-debugchecks=1`, which is always off by default, and can be disabled at compile time, allowing the code to dropped out. these checks could be slow and heavyweight. they probably wouldn't be run in travis.
 * `traditional_assert(chk)` -- evaluate for test builds, but don't evaluate in release builds

I think the `traditional_assert` style probably makes test and release builds different in ways that are hard to ensure are correct (in Haskell or Rust maybe we could ensure that chk is side effect free via the type-system, but that doesn't seem plausible in C++), so I don't think that's a good idea.

I could see any of the others being reasonable though.

MarcoFalke commented on src/check.h:
```diff
@@ -0,0 +1,66 @@
+#ifdef ABORT_ON_FAILED_DCHECK
+#define DCHECK_FAILURE(expression, file, line, func)              \
+    do {                                                          \
+        tfm::format(std::cerr, "%s:%d: %s: DCHECK(%s) failed.\n", \
```
```suggestion
        tfm::format(std::cerr, "%s:%d: %s: DCHECK(%s) failed. You may report this issue here: %s.\n", \
        ...
        PACKAGE_BUGREPORT)
```

practicalswift replied:
Added!

MarcoFalke commented:
Why was this closed?

practicalswift commented:
Rebased and renamed back from `DCHECK` to the originally suggested name `CHECK` (as first suggested by @sipa in #4576).

I find `DCHECK` somewhat unintuitive: I prefer `CHECK` or `ASSUME`. Let's bikeshed! :)

ryanofsky commented:
> Rebased and renamed back from `DCHECK` to the originally suggested name `CHECK` (as first suggested by @sipa in #4576).
> 
> I find `DCHECK` somewhat unintuitive: I prefer `CHECK` or `ASSUME`. Let's bikeshed! :)

Not interested in bikeshedding and this is fine if other reviewers want it, but I'll just point out that as of b1eba5575b53dda9bb7765701d4a6305d07faa74 this PR makes `assert` and `CHECK` mean the exact opposite things they mean in other C++ codebases, where `CHECK` aborts in all builds and `assert` aborts in debug builds only (examples in table above https://github.com/bitcoin/bitcoin/pull/16136#issuecomment-501779934).

A good way to provide more clarity, I think, would be to follow convention used in other libraries and add a pair of `CHECK`/`DCHECK` macros instead of a single macro, avoiding use of any assert. An advantage of doing this is that both macros could do extra things like log to `debug.log` or capture stack traces before aborting, which aren't so easily possible with `assert`.

practicalswift commented:
@ryanofsky What about `ASSUME`? :)

ryanofsky commented:
> @ryanofsky What about `ASSUME`? :)

Seems fine, and I personally don't care about the name. I'm trying to say that if you add a macro that aborts and reports errors in debug builds only, it would be good to have a corresponding macro that does the same thing in all builds, and that naming I've seen before for this is CHECK/DCHECK.

ajtowns commented on src/assume.h:
```diff
@@ -0,0 +1,62 @@
+// More specifically:
+//
+// * If compiled with -DABORT_ON_FAILED_ASSUME (set by --enable-debug):
+//   Evaluate expression and abort if expression is falsy.
```
"false"

ajtowns commented on src/assume.h:
```diff
@@ -0,0 +1,62 @@
+    } while (false)
+#else
+#define ASSUME_FAILED(expression, file, line, func) ((void)0)
+#endif
```
I'm still not sure it wouldn't be better unconditionally log (independent of compile flags) something like `%s:%d: ASSUME(%s) failed` and change the testing harnesses to emit a failure if that happens? If we continue fine when running release versions anyway, wouldn't it make more sense to capture all the failures while testing to reduce the number of restarts needed?

My inclination would be more along the lines of:

```
if (!(expression)) {
    LogPrintf("Assumption failed %s:%d: %s", __FILE__,__LINE__, #expression);
    if (!Params().IsTestChain()) std::abort();
}
```

(or have the `std::abort()` always happen if a compile time flag is set, if that's helpful for static analysis)

MarcoFalke replied:
has been fixed, I believe

ajtowns commented on src/assume.h:
```diff
@@ -0,0 +1,62 @@
+// Bitcoin Core is always compiled with assertions enabled. assert(...)
+// is thus only appropriate for documenting critical assumptions
+// where a failed assumption should lead to immediate abortion also
+// in production environments.
```
It's not really clear to me when you should choose `assert`  vs `ASSUME` -- neither is any quicker because both always run the test, and if the assumptions you had when writing the code were wrong, how confident can you really be that continuing won't cause horrible corruption to decide that continuing is okay?

ajtowns reviewed (commented):
I like this, but am not sure it makes sense :)

MarcoFalke commented:
Should be moved to the existing `util/check.h`

practicalswift commented:
@MarcoFalke Good idea: addressed! Please re-review.

MarcoFalke commented:
Would it make sense to add a wrapper around our current `assert`, which always aborts when false (even in production, with debug disabled)?

We have vague error reports like #17298, that obviously lack details in the debug log, because logging wasn't enabled. Maybe an wrapper around our current assert that still aborts the program, but also dumps a full `-debug=1` log for the last `n` (lets say 10 or 100) lines of debug log?

practicalswift commented:
@MarcoFalke That could be nifty to have. I would like to keep this PR and `ASSUME(…)` as simple as possible, so I'll leave that for a future PR.

My suggestion is that we start with something super simple in and then we can iterate from that as we gain experience with the different use cases :)

MarcoFalke commented on src/util/check.h:
```diff
@@ -38,4 +39,59 @@ class NonFatalCheckError : public std::runtime_error
+#define ASSUME_FAILED(expression, file, line, func)                                                  \
+    do {                                                                                             \
+        tfm::format(std::cerr, "%s:%d: %s: ASSUME(%s) failed. You may report this issue here: %s\n", \
+            file, line, func, expression, PACKAGE_BUGREPORT);                                        \
```
Shouldn't this be logged even if the debug flag is not set at compile time?

practicalswift replied:
That's not how I intended it to work, but now that you suggest it there is perhaps no downside in logging it also in non-debug/non-fuzz mode? Since we're only writing to stderr (and not to disk) it should be safe. I'll update.

MarcoFalke commented:
> @MarcoFalke That could be nifty to have. I would like to keep this PR and ASSUME(…) as simple as possible, so I'll leave that for a future PR.

Fine, but at the very least we should decide on naming. I guess, given that this patch does not use the `CHECK`/`DCHECK` naming, your proposed naming would be `ASSERT`/`ASSUME`?

practicalswift commented:
@MarcoFalke I think `ASSERT` and `ASSUME` are fine, but I'm happy to change that to whatever the consensus opinion is :)

I've encountered a few places where `ASSUME` would be handy when fuzzing: places where I want the program to crash when running under a fuzzer but where we certainly don't want to abort in production. So for me getting the macro in is the important thing -- I can live with any name :)

MarcoFalke commented:
ACK 674f9d59ba4c1a1d4662e79467cc417a2988cf15, didn't look at the configure changes

practicalswift commented:
Closing due to lack of interest :)

ajtowns commented:
We have `CHECK_NONFATAL(cond)` in `util/check.h` now, which is always compiled, always evaluated, but only throws an exception rather than aborting the program. Probably best to see how that works for a while

MarcoFalke commented:
I think `ASSUME` serves a different use case than `CHECK_NONFATAL`, but it might be best to introduce it when it is actually used in the code.

MarcoFalke commented:
Revived in #20138 with an actual use-case (I hope)

hebasto commented:
> Revived in #20138 with an actual use-case (I hope)

Actually,  #20255 :)
